test/e2e/serial/tests/workload_placement.go

```diff
@@ -371,6 +386,13 @@ var _ = Describe("[serial][disruptive][scheduler] numaresources workload placeme
 				return fxt.Client.Update(context.TODO(), updatedDp)
 			}).WithTimeout(2 * time.Minute).WithPolling(10 * time.Second).ShouldNot(HaveOccurred())
 
+			Eventually(func() int {
+				r, err := replicasets.GetReplicasetByDeployment(fxt.Client, context.TODO(), *updatedDp)
+				Expect(err).ToNot(HaveOccurred())
+				klog.Infof("number of replicasets under deployment %q is %d", namespacedDpName, len(r))
+				return len(r)
+			}).WithTimeout(10*time.Second).WithPolling(1*time.Second).Should(Equal(len(dpReplicas)+1), "deployment %q replicasets were not increased by 1", namespacedDpName)
```

Comment:
likewise, I'm missing what this check adds that the previous check right below for the deployment to complete is neglecting. It seems we missed to check a key detail, but which one?

Reply:
once I have a cluster available I'll add some debugging log like replica number under the deployment to validate that the test failed because missing the new replica

Reply:
so after adding more logs to examine the replicas under the deployment, my guess is justified because we see this:

```
I0821 09:03:37.771974  247702 deployment.go:42] deployment e2e-test-workload-placement-6409/testdp47591 complete
STEP: replicasets number is 2 should be 3 08/21/23 09:03:37.782
```

and the debugging code was added after verifying the dp was complete. 
this means we check the dp is complete "too early" -> earlier than having the new replicaset created.

Reply:
thanks, this indeed proves you found something which is worth a fix. I think we should double down and understand why the existing check is not good enough, and fix that. If this proves to be impossible, or too hard, then we should add and additional check. But in general I'd expect the existing check to cover this case. It doesn't and this means the existing code is buggy, or incomplete.

Reply:
ForDeploymentComplete only checks if the pod under the deployment is ready. In the context of this test, we change the resources of the container resulting in creating new replicaset -> recreating pod. 
I tried verifying that the DP is indeed updated (fetching the current dp and comparing it with the new expected spec) but that didn't help. 
From what I see manually, and using the automation, ForDeploymentComplete will always report success in this context because the deployment is not new, but updated. Simplest example to justify that is verifying the case where the new pod is pending, and it turns that in this case the deployment will still be considered as complete because the old pod is still there and running.

```
[root@registry ~]# oc get pod,deployment,replicaset
NAME                               READY   STATUS    RESTARTS   AGE
pod/padder-2588                    1/1     Running   0          59s
pod/padder-6902                    1/1     Running   0          59s
pod/testdp47591-586c9f5494-cngvq   0/2     Pending   0          12s
pod/testdp47591-5b9d4d78fd-qxspf   2/2     Running   0          2m16s
NAME                          READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/testdp47591   1/1     1            1           3m15s
NAME                                     DESIRED   CURRENT   READY   AGE
replicaset.apps/testdp47591-586c9f5494   1         1         0       12s
replicaset.apps/testdp47591-5b9d4d78fd   1         1         1       2m16s
replicaset.apps/testdp47591-64894fcd94   0         0         0       3m15s
```

Because of this, the deployment complete check is not enough for this test and we need to check what happens in between the checks boiling down to verify the replicaset number is actually increased. From there we guarantee either a success or a clear failure that will be caught in later stages (e.g having 2 pods ,one in pending/failing state). Thus I believe we need to have this intermediate check. I have added clarification regarding how deployment readiness is decided so it will be taken into consideration once used.

Reply:
The ForDeploymentComplete function is used in many places, but as of now, there is no obvious flakiness caused in other tests, and if there will be I believe it best be handled by adding additional checks (before/after the function call) as the context of the test requires.

Reply:
ok, thanks for the extra investigation. This justifies the extra code specific to this test.
At the same time, I can't help but notice there's a disconnect in what we expect (and how we use) deployments and the way they actually behave. We will need to handle this somehow in the future.

Reply:
I agree there is room for improvement in this regards. I have checked other uses of this function, and I can truly tell that if there wasn't the wait-for-NRT-settle and if the test had a conflicting expectation to the old deployment configuration, then there would be a good chance the test may fail due to this same disconnection.